src/Core/maps/src/Platform/iOS/MauiMKMapView.cs

```diff
@@ -0,0 +1,305 @@
+﻿using System;
+using System.Collections;
+using System.Linq;
+using CoreLocation;
+using MapKit;
+using Microsoft.Maui.Maps.Handlers;
+using Microsoft.Maui.Platform;
+using ObjCRuntime;
+using UIKit;
+
+namespace Microsoft.Maui.Maps.Platform
+{
+	internal class MauiMKMapView : MKMapView
+	{
+		IMapHandler? _handler;
+		object? _lastTouchedView;
+
+		public MauiMKMapView(IMapHandler handler)
+		{
+			_handler = handler;
+			OverlayRenderer = GetViewForOverlay1;
+		}
+
+		protected override void Dispose(bool disposing)
+		{
+			_handler = null;
+			base.Dispose(disposing);
+		}
+
+		public override void MovedToWindow()
+		{
+			base.MovedToWindow();
+			if (Window != null)
+				Startup();
+			else
+				Cleanup();
+		}
+
+		void Startup()
+		{
+			RegionChanged += MkMapViewOnRegionChanged;
+			DidSelectAnnotationView += MkMapViewOnAnnotationViewSelected;
+		}
+
+		void Cleanup()
+		{
+			if (_handler == null)
+				return;
+
+			RegionChanged -= MkMapViewOnRegionChanged;
+			DidSelectAnnotationView -= MkMapViewOnAnnotationViewSelected;
+			_handler = null;
+		}
+
+		void MkMapViewOnAnnotationViewSelected(object? sender, MKAnnotationViewEventArgs e)
+		{
+			var annotation = e.View.Annotation;
+			var pin = GetPinForAnnotation(annotation!);
+
+			if (pin != null)
+			{
+				// SendMarkerClick() returns the value of PinClickedEventArgs.HideInfoWindow
+				// Hide the info window by deselecting the annotation
+				bool deselect = pin.SendMarkerClick();
+				if (deselect)
+				{
+					DeselectAnnotation(annotation, false);
+				}
+			}
+		}
+
+		void MkMapViewOnRegionChanged(object? sender, MKMapViewChangeEventArgs e)
+		{
+			if (_handler?.VirtualView != null)
+				_handler.VirtualView.VisibleRegion = new MapSpan(new Devices.Sensors.Location(Region.Center.Latitude, Region.Center.Longitude), Region.Span.LatitudeDelta, Region.Span.LongitudeDelta);
+		}
+
+		IMapPin GetPinForAnnotation(IMKAnnotation annotation)
+		{
+			IMapPin targetPin = null!;
+			IMap map = _handler?.VirtualView!;
+
+			for (int i = 0; i < map.Pins.Count; i++)
+			{
+				var pin = map.Pins[i];
+				if ((pin?.MarkerId as IMKAnnotation) == annotation)
+				{
+					targetPin = pin;
+					break;
+				}
+			}
+
+			return targetPin;
+		}
+#pragma warning disable CS8603 // Possible null reference return.
+		protected virtual MKOverlayRenderer GetViewForOverlay1(MKMapView mapview, IMKOverlay overlay)
+		{
+			return overlay switch
+			{
+				MKPolyline polyline => GetViewForPolyline(polyline),
+				MKPolygon polygon => GetViewForPolygon(polygon),
+				MKCircle circle => GetViewForCircle(circle),
+				_ => null,
```

Comment:
Should this throw an exception? That way we can get rid of the pragma disables?

Reply:
Done